source/util/parse_number_internal.h

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2015-2016 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_UTIL_PARSE_NUMBER_INTERNAL_H_
+#define LIBSPIRV_UTIL_PARSE_NUMBER_INTERNAL_H_
+
+#include "util/parse_number.h"
+
+namespace spvutils {
+
+// A templated class with a static member function Clamp, where Clamp
+// sets a referenced value of type T to 0 if T is an unsigned
+// integer type, and returns true if it modified the referenced
+// value.
+template <typename T, typename = void>
+class ClampToZeroIfUnsignedType {
+ public:
+  // The default specialization does not clamp the value.
+  static bool Clamp(T*) { return false; }
+};
+
+// The specialization of ClampToZeroIfUnsignedType for unsigned integer
+// types.
+template <typename T>
+class ClampToZeroIfUnsignedType<
+    T, typename std::enable_if<std::is_unsigned<T>::value>::type> {
+ public:
+  static bool Clamp(T* value_pointer) {
+    if (*value_pointer) {
+      *value_pointer = 0;
+      return true;
+    }
+    return false;
+  }
+};
+
+// Returns SPV_SUCCESS if the given value fits within the target scalar
+// integral type.  The target type may have an unusual bit width.
+// If the value was originally specified as a hexadecimal number, then
+// the overflow bits should be zero.  If it was hex and the target type is
+// signed, then return the sign-extended value through the
+// updated_value_for_hex pointer argument.
+// On failure, return the given error code and emit a diagnostic if that error
+// code is not SPV_FAILED_MATCH.
+template <typename T>
+bool checkRangeAndIfHexThenSignExtend(T value, const IdType& type, bool is_hex,
```

Comment:
I'd like to see the names of various types, functions to be corrected to follow the google style guide. But definitely not in the same commit.